File: src/ormai/adapters/tortoise/adapter.py
"""
Tortoise ORM adapter implementation.

The main adapter class that implements the OrmAdapter interface for Tortoise ORM.
"""

from __future__ import annotations

from collections.abc import Callable
from typing import Any, TypeVar

from tortoise.functions import Avg, Count, Max, Min, Sum
from tortoise.transactions import in_transaction

from ormai.adapters.base import CompiledQuery, OrmAdapter
from ormai.adapters.tortoise.compiler import TortoiseCompiler
from ormai.adapters.tortoise.introspection import TortoiseIntrospector
from ormai.adapters.tortoise.mutations import MutationExecutor
from ormai.core.context import RunContext
from ormai.core.dsl import (
    AggregateRequest,
    AggregateResult,
    BulkUpdateRequest,
    BulkUpdateResult,
    CreateRequest,
    CreateResult,
    DeleteRequest,
    DeleteResult,
    GetRequest,
    GetResult,
    QueryRequest,
    QueryResult,
    UpdateRequest,
    UpdateResult,
)
from ormai.core.types import SchemaMetadata
from ormai.policy.models import Policy
from ormai.policy.redaction import Redactor

T = TypeVar("T")


class TortoiseAdapter(OrmAdapter):
    """
    Tortoise ORM adapter for OrmAI.

    Provides async database access through Tortoise ORM.
    """

    def __init__(
        self,
        models: list[type],
        policy: Policy,
        connection_name: str = "default",
    ) -> None:
        """
        Initialize the Tortoise adapter.

        Args:
            models: List of Tortoise model classes to expose
            policy: Policy configuration
            connection_name: Tortoise connection name to use
        """
        self.models = models
        self.policy = policy
        self.connection_name = connection_name

        # Initialize components
        self.introspector = TortoiseIntrospector(models)

        # Cache schema on init
        self._schema: SchemaMetadata | None = None
        self._compiler: TortoiseCompiler | None = None
        self._mutation_executor: MutationExecutor | None = None

    @property
    def schema(self) -> SchemaMetadata:
        """Get cached schema, introspecting if needed."""
        if self._schema is None:
            self._schema = self.introspector.introspect()
        return self._schema

    @property
    def compiler(self) -> TortoiseCompiler:
        """Get the query compiler."""
        if self._compiler is None:
            model_map = {m.__name__: m for m in self.models}
            self._compiler = TortoiseCompiler(model_map, self.policy, self.schema)
        return self._compiler

    @property
    def mutation_executor(self) -> MutationExecutor:
        """Get the mutation executor."""
        if self._mutation_executor is None:
            self._mutation_executor = MutationExecutor(self)
        return self._mutation_executor

    async def introspect(self) -> SchemaMetadata:
        """Introspect the database schema."""
        return self.schema

    def compile_query(
        self,
        request: QueryRequest,
        ctx: RunContext,
        policy: Policy,  # noqa: ARG002
        schema: SchemaMetadata,  # noqa: ARG002
    ) -> CompiledQuery:
        """Compile a query request."""
        return self.compiler.compile_query(request, ctx)

    def compile_get(
        self,
        request: GetRequest,
        ctx: RunContext,
        policy: Policy,  # noqa: ARG002
        schema: SchemaMetadata,  # noqa: ARG002
    ) -> CompiledQuery:
        """Compile a get request."""
        return self.compiler.compile_get(request, ctx)

    def compile_aggregate(
        self,
        request: AggregateRequest,
        ctx: RunContext,
        policy: Policy,  # noqa: ARG002
        schema: SchemaMetadata,  # noqa: ARG002
    ) -> CompiledQuery:
        """Compile an aggregate request."""
        return self.compiler.compile_aggregate(request, ctx)

    async def execute_query(
        self,
        compiled: CompiledQuery,
        ctx: RunContext,  # noqa: ARG002
    ) -> QueryResult:
        """Execute a compiled query."""
        queryset = compiled.query
        rows = await queryset

        # Convert to dicts and apply redaction
        request = compiled.request
        if not isinstance(request, QueryRequest):
            raise ValueError("Expected QueryRequest")

        data = self._rows_to_dicts(rows, compiled.select_fields, request.model)

        # Build pagination info
        has_more = len(data) >= request.take
        next_cursor = None
        if has_more:
            current_offset = self._get_current_offset(request.cursor)
            next_cursor = TortoiseCompiler.encode_cursor(current_offset + request.take)

        return QueryResult(
            data=data,
            next_cursor=next_cursor,
            has_more=has_more,
        )

    async def execute_get(
        self,
        compiled: CompiledQuery,
        ctx: RunContext,  # noqa: ARG002
    ) -> GetResult:
        """Execute a get request."""
        queryset = compiled.query
        row = await queryset.first()

        if row is None:
            return GetResult(data=None, found=False)

        request = compiled.request
        if not isinstance(request, GetRequest):
            raise ValueError("Expected GetRequest")

        data = self._row_to_dict(row, compiled.select_fields, request.model)
        return GetResult(data=data, found=True)

    async def execute_aggregate(
        self,
        compiled: CompiledQuery,
        ctx: RunContext,  # noqa: ARG002
    ) -> AggregateResult:
        """Execute an aggregation."""
        agg_info = compiled.query
        queryset = agg_info["queryset"]
        operation = agg_info["operation"]
        field = agg_info["field"]

        request = compiled.request
        if not isinstance(request, AggregateRequest):
            raise ValueError("Expected AggregateRequest")

        # Execute the aggregation
        value = await self._execute_aggregation(queryset, operation, field)

        return AggregateResult(
            value=value,
            operation=operation,
            field=field,
        )

    async def _execute_aggregation(
        self,
        queryset: Any,
        operation: str,
        field: str | None,
    ) -> Any:
        """Execute a specific aggregation operation."""
        match operation:
            case "count":
                if field:
                    result = await queryset.annotate(val=Count(field)).values("val")
                    return result[0]["val"] if result else 0
                return await queryset.count()
            case "sum":
                result = await queryset.annotate(val=Sum(field)).values("val")
                return result[0]["val"] if result else 0
            case "avg":
                result = await queryset.annotate(val=Avg(field)).values("val")
                return result[0]["val"] if result else None
            case "min":
                result = await queryset.annotate(val=Min(field)).values("val")
                return result[0]["val"] if result else None
            case "max":
                result = await queryset.annotate(val=Max(field)).values("val")
                return result[0]["val"] if result else None
            case _:
                raise ValueError(f"Unsupported aggregation: {operation}")

    async def transaction(
        self,
        ctx: RunContext,  # noqa: ARG002
        fn: Callable[..., T],
        *args: Any,
        **kwargs: Any,
    ) -> T:
        """Execute a function within a transaction."""
        async with in_transaction(self.connection_name):
            return await fn(*args, **kwargs)

    def _rows_to_dicts(
        self,
        rows: list[Any],
        fields: list[str],
        model_name: str,
    ) -> list[dict[str, Any]]:
        """Convert ORM rows to dicts with redaction."""
        model_policy = self.policy.get_model_policy(model_name)
        redactor = Redactor(model_policy) if model_policy else None

        result = []
        for row in rows:
            row_dict = self._row_to_dict(row, fields, model_name, redactor)
            result.append(row_dict)
        return result

    def _row_to_dict(
        self,
        row: Any,
        fields: list[str],
        model_name: str,
        redactor: Redactor | None = None,
    ) -> dict[str, Any]:
        """Convert a single ORM row to dict with redaction."""
        # Extract only allowed fields
        data = {}
        for field in fields:
            value = getattr(row, field, None)
            # Handle datetime serialization
            if hasattr(value, "isoformat"):
                value = value.isoformat()
            data[field] = value

        # Apply redaction if we have a policy
        if redactor is None:
            model_policy = self.policy.get_model_policy(model_name)
            if model_policy:
                redactor = Redactor(model_policy)

        if redactor:
            data = redactor.redact_record(data)

        return data

    def _get_current_offset(self, cursor: str | None) -> int:
        """Get current offset from cursor."""
        if cursor is None:
            return 0
        try:
            return int(cursor)
        except ValueError:
            return 0

    # =========================================================================
    # MUTATION METHODS
    # =========================================================================

    def compile_create(
        self,
        request: CreateRequest,
        ctx: RunContext,
        policy: Policy,
        schema: SchemaMetadata,
    ) -> CompiledQuery:
        """Compile a create request."""
        return self.mutation_executor.compile_create(request, ctx, policy, schema)

    def compile_update(
        self,
        request: UpdateRequest,
        ctx: RunContext,
        policy: Policy,
        schema: SchemaMetadata,
    ) -> CompiledQuery:
        """Compile an update request."""
        return self.mutation_executor.compile_update(request, ctx, policy, schema)

    def compile_delete(
        self,
        request: DeleteRequest,
        ctx: RunContext,
        policy: Policy,
        schema: SchemaMetadata,
    ) -> CompiledQuery:
        """Compile a delete request."""
        return self.mutation_executor.compile_delete(request, ctx, policy, schema)

    def compile_bulk_update(
        self,
        request: BulkUpdateRequest,
        ctx: RunContext,
        policy: Policy,
        schema: SchemaMetadata,
    ) -> CompiledQuery:
        """Compile a bulk update request."""
        return self.mutation_executor.compile_bulk_update(request, ctx, policy, schema)

    async def execute_create(
        self,
        compiled: CompiledQuery,
        ctx: RunContext,
    ) -> CreateResult:
        """Execute a create request."""
        return await self.mutation_executor.execute_create(compiled, ctx)

    async def execute_update(
        self,
        compiled: CompiledQuery,
        ctx: RunContext,
    ) -> UpdateResult:
        """Execute an update request."""
        return await self.mutation_executor.execute_update(compiled, ctx)

    async def execute_delete(
        self,
        compiled: CompiledQuery,
        ctx: RunContext,
    ) -> DeleteResult:
        """Execute a delete request."""
        return await self.mutation_executor.execute_delete(compiled, ctx)

    async def execute_bulk_update(
        self,
        compiled: CompiledQuery,
        ctx: RunContext,
    ) -> BulkUpdateResult:
        """Execute a bulk update request."""
        return await self.mutation_executor.execute_bulk_update(compiled, ctx)

File: ormai-ts/src/adapters/drizzle/adapter.ts
/**
 * Drizzle ORM adapter for OrmAI.
 */

import type { RunContext } from '../../core/context.js';
import type {
  AggregateRequest,
  AggregateResult,
  BulkUpdateRequest,
  BulkUpdateResult,
  CreateRequest,
  CreateResult,
  DeleteRequest,
  DeleteResult,
  GetRequest,
  GetResult,
  QueryRequest,
  QueryResult,
  UpdateRequest,
  UpdateResult,
} from '../../core/dsl.js';
import type { SchemaMetadata } from '../../core/types.js';
import type { Policy } from '../../policy/models.js';
import { PolicyEngine } from '../../policy/engine.js';
import { PolicyUtils } from '../../policy/models.js';
import type { CompiledMutation, CompiledQuery } from '../base.js';
import { BaseOrmAdapter, applyRedactionToRecords } from '../base.js';
import {
  DrizzleCompiler,
  type DrizzleOperators,
  type DrizzleTableRef,
  type CompiledDrizzleQuery,
  type CompiledDrizzleMutation,
} from './compiler.js';
import { DrizzleIntrospector, type DrizzleRelation, type DrizzleSchema } from './introspection.js';

/**
 * Drizzle aggregate function builder.
 *
 * Used to build SQL-level aggregate queries (COUNT, SUM, AVG, MIN, MAX).
 * The callback receives the table and column references and should return
 * a promise resolving to a single-row aggregate result.
 */
export type DrizzleAggregateFn = (
  table: DrizzleTableRef,
  field: string | null,
  operation: AggregateRequest['operation'],
  where?: unknown,
) => Promise<Record<string, unknown>>;

/**
 * Drizzle database instance type.
 */
export interface DrizzleDB {
  select: (fields?: Record<string, unknown>) => DrizzleQueryBuilder;
  insert: (table: DrizzleTableRef) => DrizzleInsertBuilder;
  update: (table: DrizzleTableRef) => DrizzleUpdateBuilder;
  delete: (table: DrizzleTableRef) => DrizzleDeleteBuilder;
  transaction: <T>(fn: (tx: DrizzleDB) => Promise<T>) => Promise<T>;
}

/**
 * Drizzle query builder interface.
 */
export interface DrizzleQueryBuilder {
  from: (table: DrizzleTableRef) => DrizzleQueryBuilder;
  where: (...conditions: unknown[]) => DrizzleQueryBuilder;
  orderBy: (...orders: unknown[]) => DrizzleQueryBuilder;
  limit: (n: number) => DrizzleQueryBuilder;
  offset: (n: number) => DrizzleQueryBuilder;
  then: <T>(resolve: (value: unknown[]) => T) => Promise<T>;
}

/**
 * Drizzle insert builder interface.
 */
export interface DrizzleInsertBuilder {
  values: (data: Record<string, unknown> | Record<string, unknown>[]) => DrizzleInsertBuilder;
  returning: () => DrizzleInsertBuilder;
  then: <T>(resolve: (value: unknown[]) => T) => Promise<T>;
}

/**
 * Drizzle update builder interface.
 */
export interface DrizzleUpdateBuilder {
  set: (data: Record<string, unknown>) => DrizzleUpdateBuilder;
  where: (...conditions: unknown[]) => DrizzleUpdateBuilder;
  returning: () => DrizzleUpdateBuilder;
  then: <T>(resolve: (value: unknown[]) => T) => Promise<T>;
}

/**
 * Drizzle delete builder interface.
 */
export interface DrizzleDeleteBuilder {
  where: (...conditions: unknown[]) => DrizzleDeleteBuilder;
  returning: () => DrizzleDeleteBuilder;
  then: <T>(resolve: (value: unknown[]) => T) => Promise<T>;
}

/**
 * Drizzle adapter configuration.
 */
export interface DrizzleAdapterConfig {
  /** Drizzle database instance */
  db: DrizzleDB;

  /** Drizzle schema (tables) */
  schema: DrizzleSchema;

  /** Drizzle operators */
  operators: DrizzleOperators;

  /** Drizzle relations (optional) */
  relations?: Record<string, DrizzleRelation[]>;

  /** Model names to expose (optional, defaults to all) */
  models?: string[];

  /**
   * Server-side aggregate function.
   * If provided, the adapter uses this for sum/avg/min/max operations
   * instead of fetching all rows and computing client-side.
   * For count, the adapter always uses a filtered select for efficiency.
   */
  aggregateFn?: DrizzleAggregateFn;
}

/**
 * Drizzle adapter for OrmAI.
 */
export class DrizzleAdapter extends BaseOrmAdapter<DrizzleDB, CompiledDrizzleQuery, CompiledDrizzleMutation> {
  private readonly db: DrizzleDB;
  private readonly drizzleSchema: DrizzleSchema;
  private readonly operators: DrizzleOperators;
  private readonly relations: Record<string, DrizzleRelation[]>;
  private readonly modelFilter?: string[];
  private readonly aggregateFn?: DrizzleAggregateFn;
  private readonly compiler: DrizzleCompiler;
  private cachedSchema?: SchemaMetadata;

  constructor(config: DrizzleAdapterConfig) {
    super();
    this.db = config.db;
    this.drizzleSchema = config.schema;
    this.operators = config.operators;
    this.relations = config.relations ?? {};
    this.modelFilter = config.models;
    this.aggregateFn = config.aggregateFn;

    // Build table map from schema
    const tables: Record<string, DrizzleTableRef> = {};
    for (const [name, table] of Object.entries(this.drizzleSchema)) {
      tables[name] = table as DrizzleTableRef;
      if (table._ && table._.name) {
        tables[table._.name as string] = table as DrizzleTableRef;
      }
    }

    this.compiler = new DrizzleCompiler(tables, this.operators);
  }

  /**
   * Introspect the Drizzle schema.
   */
  async introspect(): Promise<SchemaMetadata> {
    if (this.cachedSchema) {
      return this.cachedSchema;
    }

    const introspector = new DrizzleIntrospector(this.drizzleSchema, this.relations);
    let schema = introspector.introspect();

    // Filter models if specified
    if (this.modelFilter) {
      const filtered: Record<string, typeof schema.models[string]> = {};
      for (const name of this.modelFilter) {
        if (schema.models[name]) {
          filtered[name] = schema.models[name];
        }
      }
      schema = { models: filtered };
    }

    this.cachedSchema = schema;
    return schema;
  }

  /**
   * Compile a query request.
   */
  compileQuery(
    request: QueryRequest,
    ctx: RunContext<DrizzleDB>,
    policy: Policy,
    schema: SchemaMetadata
  ): CompiledQuery<CompiledDrizzleQuery> {
    const engine = new PolicyEngine(policy, schema);
    const decision = engine.validateQuery(request, ctx);

    const table = this.compiler.getTable(request.model);

    // Merge user filters with scope filters
    const allFilters = [
      ...(request.where ?? []),
      ...decision.injectedFilters,
    ];

    const compiled: CompiledDrizzleQuery = {
      table,
      where: allFilters.length > 0 ? this.compiler.compileFilters(allFilters, table) : undefined,
      orderBy: request.orderBy ? this.compiler.compileOrderBy(request.orderBy, table) : [],
      limit: request.take,
      offset: undefined,
      columns: decision.allowedFields.length > 0
        ? this.compiler.compileSelect(decision.allowedFields, table)
        : undefined,
    };

    return {
      query: compiled,
      request,
      selectFields: decision.allowedFields,
      injectedFilters: decision.injectedFilters,
      policyDecisions: decision.decisions,
      timeoutMs: decision.budget?.statementTimeoutMs,
      redactionRules: Object.keys(decision.redactionRules).length > 0
        ? decision.redactionRules
        : undefined,
    };
  }

  /**
   * Execute a compiled query.
   */
  async executeQuery(
    compiled: CompiledQuery<CompiledDrizzleQuery>,
    ctx: RunContext<DrizzleDB>
  ): Promise<QueryResult> {
    const { query } = compiled;
    const db = ctx.db ?? this.db;
    const request = compiled.request as QueryRequest;

    try {
      let builder = db.select().from(query.table);

      if (query.where) {
        builder = builder.where(query.where);
      }

      if (query.orderBy && query.orderBy.length > 0) {
        builder = builder.orderBy(...query.orderBy);
      }

      if (query.limit !== undefined) {
        builder = builder.limit(query.limit);
      }

      if (query.offset !== undefined) {
        builder = builder.offset(query.offset);
      }

      const rows = await builder as Record<string, unknown>[];

      // Apply redaction rules
      const redactedRows = applyRedactionToRecords(rows, compiled.redactionRules);

      return {
        data: redactedRows,
        nextCursor: null,
        hasMore: rows.length >= request.take,
        totalCount: null,
      };
    } catch (error) {
      throw new Error(`Query execution failed: ${(error as Error).message}`);
    }
  }

  /**
   * Compile a get request.
   */
  compileGet(
    request: GetRequest,
    ctx: RunContext<DrizzleDB>,
    policy: Policy,
    schema: SchemaMetadata
  ): CompiledQuery<CompiledDrizzleQuery> {
    const engine = new PolicyEngine(policy, schema);
    const decision = engine.validateGet(request, ctx);

    const table = this.compiler.getTable(request.model);
    const model = schema.models[request.model];

    if (!model) {
      throw new Error(`Model '${request.model}' not found in schema`);
    }

    // Build primary key filter
    const pkFilter = { field: model.primaryKey, op: 'eq' as const, value: request.id };
    const allFilters = [pkFilter, ...decision.injectedFilters];

    const compiled: CompiledDrizzleQuery = {
      table,
      where: this.compiler.compileFilters(allFilters, table),
      orderBy: [],
      limit: 1,
      columns: decision.allowedFields.length > 0
        ? this.compiler.compileSelect(decision.allowedFields, table)
        : undefined,
    };

    return {
      query: compiled,
      request,
      selectFields: decision.allowedFields,
      injectedFilters: decision.injectedFilters,
      policyDecisions: decision.decisions,
      timeoutMs: decision.budget?.statementTimeoutMs,
      redactionRules: Object.keys(decision.redactionRules).length > 0
        ? decision.redactionRules
        : undefined,
    };
  }

  /**
   * Execute a compiled get query.
   */
  async executeGet(
    compiled: CompiledQuery<CompiledDrizzleQuery>,
    ctx: RunContext<DrizzleDB>
  ): Promise<GetResult> {
    const result = await this.executeQuery(compiled, ctx);
    const row = result.data[0] ?? null;
    return { data: row, found: row !== null };
  }

  /**
   * Compile an aggregate request.
   */
  compileAggregate(
    request: AggregateRequest,
    ctx: RunContext<DrizzleDB>,
    policy: Policy,
    schema: SchemaMetadata
  ): CompiledQuery<CompiledDrizzleQuery> {
    const engine = new PolicyEngine(policy, schema);
    const decision = engine.validateAggregate(request, ctx);

    const table = this.compiler.getTable(request.model);

    // Merge user filters with scope filters
    const allFilters = [
      ...(request.where ?? []),
      ...decision.injectedFilters,
    ];

    const compiled: CompiledDrizzleQuery = {
      table,
      where: allFilters.length > 0 ? this.compiler.compileFilters(allFilters, table) : undefined,
      orderBy: [],
    };

    return {
      query: compiled,
      request,
      selectFields: [],
      injectedFilters: decision.injectedFilters,
      policyDecisions: decision.decisions,
      timeoutMs: decision.budget?.statementTimeoutMs,
    };
  }

  /**
   * Execute a compiled aggregate query.
   */
  async executeAggregate(
    compiled: CompiledQuery<CompiledDrizzleQuery>,
    ctx: RunContext<DrizzleDB>
  ): Promise<AggregateResult> {
    const { query } = compiled;
    const request = compiled.request as AggregateRequest;
    const db = ctx.db ?? this.db;

    try {
      // For count, use a filtered query and get row count efficiently
      if (request.operation === 'count') {
        let builder = db.select().from(query.table);
        if (query.where) {
          builder = builder.where(query.where);
        }
        // If the select builder supports a count-like approach, use that;
        // otherwise fall back to fetching rows for the count
        const rows = await builder as Record<string, unknown>[];
        return {
          value: rows.length,
          operation: 'count',
          field: request.field ?? null,
          rowCount: rows.length,
        };
      }

      // For sum/avg/min/max, use server-side aggregation if aggregateFn is provided
      if (this.aggregateFn) {
        const result = await this.aggregateFn(
          query.table,
          request.field ?? null,
          request.operation,
          query.where,
        );
        const value = result['agg_value'] ?? result['value'] ?? null;
        let rowCount = 0;
        const rowCountVal = result['row_count'] ?? result['rowCount'];
        if (typeof rowCountVal === 'number') {
          rowCount = rowCountVal;
        }
        return {
          value: value,
          operation: request.operation,
          field: request.field ?? null,
          rowCount,
        };
      }

      // Fallback: client-side aggregation (fetches all rows)
      // NOTE: Provide aggregateFn in config for production use to avoid
      // fetching entire result sets for sum/avg/min/max operations.
      let builder = db.select().from(query.table);
      if (query.where) {
        builder = builder.where(query.where);
      }
      const rows = await builder as Record<string, unknown>[];

      let value: unknown = null;

      switch (request.operation) {
        case 'sum':
          if (request.field) {
            value = rows.reduce((sum, row) => sum + (Number(row[request.field!]) || 0), 0);
          }
          break;
        case 'avg':
          if (request.field && rows.length > 0) {
            value = rows.reduce((sum, row) => sum + (Number(row[request.field!]) || 0), 0) / rows.length;
          }
          break;
        case 'min':
          if (request.field && rows.length > 0) {
            value = rows.reduce(
              (min, row) => {
                const v = Number(row[request.field!]) || 0;
                return v < min ? v : min;
              },
              Number(rows[0][request.field!]) || 0,
            );
          }
          break;
        case 'max':
          if (request.field && rows.length > 0) {
            value = rows.reduce(
              (max, row) => {
                const v = Number(row[request.field!]) || 0;
                return v > max ? v : max;
              },
              Number(rows[0][request.field!]) || 0,
            );
          }
          break;
      }

      return {
        value,
        operation: request.operation,
        field: request.field ?? null,
        rowCount: rows.length,
      };
    } catch (error) {
      throw new Error(`Aggregate execution failed: ${(error as Error).message}`);
    }
  }

  /**
   * Compile a create request.
   */
  compileCreate(
    request: CreateRequest,
    ctx: RunContext<DrizzleDB>,
    policy: Policy,
    schema: SchemaMetadata
  ): CompiledMutation<CompiledDrizzleMutation> {
    const engine = new PolicyEngine(policy, schema);
    const decision = engine.validateCreate(request, ctx);

    // Build data with scope injection
    const data = { ...request.data };
    for (const filter of decision.injectedFilters) {
      if (filter.op === 'eq') {
        data[filter.field] = filter.value;
      }
    }

    const table = this.compiler.getTable(request.model);
    const compiled: CompiledDrizzleMutation = {
      table,
      type: 'insert',
      data,
      returning: true,
    };

    return {
      mutation: compiled,
      request,
      data,
      injectedFilters: decision.injectedFilters,
      policyDecisions: decision.decisions,
      returnFields: decision.allowedFields,
    };
  }

  /**
   * Execute a compiled create mutation.
   */
  async executeCreate(
    compiled: CompiledMutation<CompiledDrizzleMutation>,
    ctx: RunContext<DrizzleDB>
  ): Promise<CreateResult> {
    const { mutation } = compiled;
    const db = ctx.db ?? this.db;

    try {
      const result = await db
        .insert(mutation.table)
        .values(mutation.data!)
        .returning() as Record<string, unknown>[];

      const record = result[0] ?? null;

      return {
        data: record,
        id: record ? record['id'] : undefined,
        success: true,
      };
    } catch (error) {
      throw new Error(`Create failed: ${(error as Error).message}`);
    }
  }

  /**
   * Compile an update request.
   */
  compileUpdate(
    request: UpdateRequest,
    ctx: RunContext<DrizzleDB>,
    policy: Policy,
    schema: SchemaMetadata
  ): CompiledMutation<CompiledDrizzleMutation> {
    const engine = new PolicyEngine(policy, schema);
    const decision = engine.validateUpdate(request, ctx);

    const table = this.compiler.getTable(request.model);
    const model = schema.models[request.model];

    if (!model) {
      throw new Error(`Model '${request.model}' not found in schema`);
    }

    // Build primary key filter with scope filters
    const pkFilter = { field: model.primaryKey, op: 'eq' as const, value: request.id };
    const allFilters = [pkFilter, ...decision.injectedFilters];

    const compiled: CompiledDrizzleMutation = {
      table,
      type: 'update',
      data: request.data,
      where: this.compiler.compileFilters(allFilters, table),
      returning: true,
    };

    return {
      mutation: compiled,
      request,
      data: request.data,
      injectedFilters: decision.injectedFilters,
      policyDecisions: decision.decisions,
      returnFields: decision.allowedFields,
    };
  }

  /**
   * Execute a compiled update mutation.
   */
  async executeUpdate(
    compiled: CompiledMutation<CompiledDrizzleMutation>,
    ctx: RunContext<DrizzleDB>
  ): Promise<UpdateResult> {
    const { mutation } = compiled;
    const db = ctx.db ?? this.db;

    try {
      let builder = db.update(mutation.table).set(mutation.data!);

      if (mutation.where) {
        builder = builder.where(mutation.where);
      }

      const result = await builder.returning() as Record<string, unknown>[];
      const record = result[0] ?? null;

      return {
        data: record,
        success: result.length > 0,
        found: result.length > 0,
      };
    } catch (error) {
      throw new Error(`Update failed: ${(error as Error).message}`);
    }
  }

  /**
   * Compile a delete request.
   */
  compileDelete(
    request: DeleteRequest,
    ctx: RunContext<DrizzleDB>,
    policy: Policy,
    schema: SchemaMetadata
  ): CompiledMutation<CompiledDrizzleMutation> {
    const engine = new PolicyEngine(policy, schema);
    const decision = engine.validateDelete(request, ctx);

    const table = this.compiler.getTable(request.model);
    const model = schema.models[request.model];

    if (!model) {
      throw new Error(`Model '${request.model}' not found in schema`);
    }

    // Build primary key filter with scope filters
    const pkFilter = { field: model.primaryKey, op: 'eq' as const, value: request.id };
    const allFilters = [pkFilter, ...decision.injectedFilters];

    const rowPolicy = PolicyUtils.getRowPolicy(policy, request.model);
    const softDelete = !request.hard && rowPolicy.softDeleteField;

    if (softDelete) {
      // Soft delete is an update
      const compiled: CompiledDrizzleMutation = {
        table,
        type: 'update',
        data: { [rowPolicy.softDeleteField!]: new Date() },
        where: this.compiler.compileFilters(allFilters, table),
        returning: true,
      };

      return {
        mutation: compiled,
        request,
        data: { [rowPolicy.softDeleteField!]: new Date() },
        injectedFilters: decision.injectedFilters,
        policyDecisions: [...decision.decisions, 'Using soft delete'],
        returnFields: [],
      };
    }

    const compiled: CompiledDrizzleMutation = {
      table,
      type: 'delete',
      where: this.compiler.compileFilters(allFilters, table),
      returning: true,
    };

    return {
      mutation: compiled,
      request,
      injectedFilters: decision.injectedFilters,
      policyDecisions: decision.decisions,
      returnFields: [],
    };
  }

  /**
   * Execute a compiled delete mutation.
   */
  async executeDelete(
    compiled: CompiledMutation<CompiledDrizzleMutation>,
    ctx: RunContext<DrizzleDB>
  ): Promise<DeleteResult> {
    const { mutation } = compiled;
    const db = ctx.db ?? this.db;

    try {
      if (mutation.type === 'update') {
        // Soft delete
        let builder = db.update(mutation.table).set(mutation.data!);
        if (mutation.where) {
          builder = builder.where(mutation.where);
        }
        const result = await builder.returning() as unknown[];
        return {
          success: result.length > 0,
          found: result.length > 0,
          softDeleted: true,
        };
      } else {
        // Hard delete
        let builder = db.delete(mutation.table);
        if (mutation.where) {
          builder = builder.where(mutation.where);
        }
        const result = await builder.returning() as unknown[];
        return {
          success: result.length > 0,
          found: result.length > 0,
          softDeleted: false,
        };
      }
    } catch (error) {
      throw new Error(`Delete failed: ${(error as Error).message}`);
    }
  }

  /**
   * Compile a bulk update request.
   */
  compileBulkUpdate(
    request: BulkUpdateRequest,
    ctx: RunContext<DrizzleDB>,
    policy: Policy,
    schema: SchemaMetadata
  ): CompiledMutation<CompiledDrizzleMutation> {
    const engine = new PolicyEngine(policy, schema);
    const decision = engine.validateBulkUpdate(request, ctx);

    const table = this.compiler.getTable(request.model);
    const model = schema.models[request.model];

    if (!model) {
      throw new Error(`Model '${request.model}' not found in schema`);
    }

    // Build where clause with all IDs and scope filters
    const idsFilter = { field: model.primaryKey, op: 'in' as const, value: request.ids };
    const allFilters = [idsFilter, ...decision.injectedFilters];

    const compiled: CompiledDrizzleMutation = {
      table,
      type: 'update',
      data: request.data,
      where: this.compiler.compileFilters(allFilters, table),
      returning: true,
    };

    return {
      mutation: compiled,
      request,
      data: request.data,
      injectedFilters: decision.injectedFilters,
      policyDecisions: decision.decisions,
      returnFields: [],
    };
  }

  /**
   * Execute a compiled bulk update mutation.
   */
  async executeBulkUpdate(
    compiled: CompiledMutation<CompiledDrizzleMutation>,
    ctx: RunContext<DrizzleDB>
  ): Promise<BulkUpdateResult> {
    const { mutation } = compiled;
    const db = ctx.db ?? this.db;

    try {
      let builder = db.update(mutation.table).set(mutation.data!);
      if (mutation.where) {
        builder = builder.where(mutation.where);
      }
      const result = await builder.returning() as unknown[];

      return {
        updatedCount: result.length,
        success: true,
        failedIds: [],
      };
    } catch (error) {
      throw new Error(`Bulk update failed: ${(error as Error).message}`);
    }
  }

  /**
   * Execute operations in a transaction.
   */
  async transaction<T>(
    ctx: RunContext<DrizzleDB>,
    fn: () => Promise<T>
  ): Promise<T> {
    const db = ctx.db ?? this.db;
    return db.transaction(async () => fn());
  }
}

/**
 * Create a Drizzle adapter.
 */
export function createDrizzleAdapter(config: DrizzleAdapterConfig): DrizzleAdapter {
  return new DrizzleAdapter(config);
}
